src/xmipp/libraries/reconstruction/subtract_projection.cpp

```diff
@@ -0,0 +1,362 @@
+	<< "Mask:\t" << fnMask      << std::endl
+	<< "Sigma:\t" << sigma       << std::endl
+	<< "Iterations:\t" << iter        << std::endl
+	<< "Cutoff frequency:\t" << cutFreq     << std::endl
+	<< "Relaxation factor:\t" << lambda      << std::endl
+	<< "Output particles:\t" << fnOut 	     << std::endl;
+ }
+
+ // usage ===================================================================
+ void ProgSubtractProjection::defineParams()
+ {
+     //Usage
+     addUsageLine("");
+     //Parameters
+     addParamsLine("-i <particles>\t: Particles metadata (.xmd file)");
+     addParamsLine("--ref <volume>\t: Reference volume to subtract");
+     addParamsLine("[-o <structure=\"\">]\t: Output filename suffix for subtracted particles");
+     addParamsLine("\t: If no name is given, then output_particles.xmd");
+     addParamsLine("[--maskVol <maskVol=\"\">]\t: 3D mask for input volume");
+     addParamsLine("[--mask <mask=\"\">]\t: 3D mask for the region of subtraction");
+     addParamsLine("[--sigma <s=3>]\t: Decay of the filter (sigma) to smooth the mask transition");
+     addParamsLine("[--iter <n=1>]\t: Number of iterations");
+     addParamsLine("[--cutFreq <f=0>]\t: Cutoff frequency (<0.5)");
+     addParamsLine("[--lambda <l=0>]\t: Relaxation factor for Fourier Amplitude POCS (between 0 and 1)");
+	 addParamsLine("[--saveProj <structure=\"\"> ]\t: Save subtraction intermediate files (projection adjusted)");
+     addExampleLine("A typical use is:",false);
+     addExampleLine("xmipp_subtract_projection -i input_particles.xmd --ref input_map.mrc --maskVol mask_vol.vol --mask mask.vol -o output_particles --iter 5 --lambda 1 --cutFreq 0.44 --sigma 3");
+ }
+
+ void POCSmaskProj(const MultidimArray<double> &mask, MultidimArray<double> &I) {
+ 	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(I)
+ 	DIRECT_MULTIDIM_ELEM(I,n)*=DIRECT_MULTIDIM_ELEM(mask,n);
+ }
+
+ void POCSFourierAmplitudeProj(const MultidimArray<double> &A, MultidimArray< std::complex<double> > &FI, double lambda, const MultidimArray<double> &rQ, int Isize) {
+ 	int Isize2 = Isize/2;
+ 	double Isizei = 1.0/Isize;
+ 	double wx;
+ 	double wy;
+ 	for (int i=0; i<YSIZE(A); ++i) {
+ 		FFT_IDX2DIGFREQ_FAST(i,Isize,Isize2,Isizei,wy)
+ 		double wy2 = wy*wy;
+ 		for (int j=0; j<XSIZE(A); ++j) {
+ 			FFT_IDX2DIGFREQ_FAST(j,Isize,Isize2,Isizei,wx)
+ 			double w = sqrt(wx*wx + wy2);
+ 			auto iw = (int)round(w*Isize);
+ 			double mod = std::abs(DIRECT_A2D_ELEM(FI,i,j));
+ 			if (mod>1e-6)
+ 				DIRECT_A2D_ELEM(FI,i,j)*=(((1-lambda)+lambda*DIRECT_A2D_ELEM(A,i,j))/mod)*DIRECT_MULTIDIM_ELEM(rQ,iw);
+ 		}
+ 	}
+ }
+
+ void POCSMinMaxProj(MultidimArray<double> &P, double Im, double IM) {
+ 	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(P) {
+ 		double val = DIRECT_MULTIDIM_ELEM(P,n);
+ 		if (val<Im)
+ 			DIRECT_MULTIDIM_ELEM(P,n) = Im;
+ 		else if (val>IM)
+ 			DIRECT_MULTIDIM_ELEM(P,n) = IM;
+ 		}
+ }
+
+ void extractPhaseProj(MultidimArray< std::complex<double> > &FI) {
```

Comment:
Do you expect this and other global methods to be used by somebody else?
If not, consider adding them to the ProgSubtractProjection class.